iavl_orphaning_tree.go

```diff
@@ -0,0 +1,108 @@
+package iavl
+
+import (
+	cmn "github.com/tendermint/tmlibs/common"
+)
+
+// orphaningTree is a tree which keeps track of orphaned nodes.
+type orphaningTree struct {
+	*IAVLTree
+
+	// A map of orphan hash to orphan version.
+	// The version stored here is the one at which the orphan's lifetime
+	// begins.
+	orphans map[string]uint64
+}
+
+// newOrphaningTree creates a new orphaning tree from the given *IAVLTree.
+func newOrphaningTree(t *IAVLTree) *orphaningTree {
+	return &orphaningTree{
+		IAVLTree: t,
+		orphans:  map[string]uint64{},
+	}
+}
+
+// Set a key on the underlying tree while storing the orphaned nodes.
+func (tree *orphaningTree) Set(key, value []byte) bool {
+	orphaned, updated := tree.IAVLTree.set(key, value)
+	tree.addOrphans(orphaned)
+	return updated
+}
+
+// Remove a key from the underlying tree while storing the orphaned nodes.
+func (tree *orphaningTree) Remove(key []byte) ([]byte, bool) {
+	val, orphaned, removed := tree.IAVLTree.remove(key)
+	tree.addOrphans(orphaned)
+	return val, removed
+}
+
+// Clone creates a clone of the tree.
+func (tree *orphaningTree) Clone() *orphaningTree {
+	inner := &IAVLTree{
+		root: tree.IAVLTree.root,
+		ndb:  tree.IAVLTree.ndb,
+	}
+	return &orphaningTree{
+		IAVLTree: inner,
+		orphans:  map[string]uint64{},
+	}
+}
+
+// Load the tree from disk, from the given root hash, including all orphans.
+func (tree *orphaningTree) Load(root []byte) {
+	if len(root) == 0 {
+		tree.root = nil
+	} else {
+		tree.root = tree.ndb.GetNode(root)
+	}
+
+	// Load orphans.
+	tree.ndb.traverseOrphansVersion(tree.root.version, func(k, v []byte) {
```

Comment:
won't this panic if `tree.root == nil` ?

Reply:
Good point. 1126fcc